### PR TITLE
Adds `pushChanges` queries for sync customization

### PR DESCRIPTION
**Pull request by sidferreira**
Based on #1027

The suggestion is to add extra parameters to `pushChanges` so people can customize the sync process.

> Note: It required more changes than I expected and I'm not sure I'm happy with it.

### CODE REVIEW

File: src/Collection/index.js [hunk 1]
```diff
@@ -83,6 +83,10 @@ export default class Collection<Record: Model> {
     return new Query(this, clauses)
   }
 
+  queryWithDeleted(...clauses: Clause[]): Query<Record> {
```
Comment (radex) [inline src/Collection/index.js:86]: what legitimate use cases are there for fetching deleted records?

**Comment by radex** [comment]
@sidferreira Can you elaborate how this new API is supposed to be used - how users would be to use it exactly? If there really is a large legitimate use case for many different ways of customizing sync queries, fine, but because of performance and correctness concerns I mentioned above, I'm hesitant to add a generic extension point like this if something simpler can solve the issue.

What I understood from @osvaldokalvaitir in #1027 is that there exists a need to make certain tables non-synchronizable (which can be done in a simpler way) and certain records created-but-not-yet-ready-to-synchronize, which also sounds like a legit improvement to Watermelon Sync that could be achieved by formalizing a new `draft` sync status without a generic customization point.

WDYT?

**Comment by osvaldokalvaitir** [comment]
> @sidferreira Can you elaborate how this new API is supposed to be used - how users would be to use it exactly? If there really is a large legitimate use case for many different ways of customizing sync queries, fine, but because of performance and correctness concerns I mentioned above, I'm hesitant to add a generic extension point like this if something simpler can solve the issue.
> 
> What I understood from @osvaldokalvaitir in #1027 is that there exists a need to make certain tables non-synchronizable (which can be done in a simpler way) and certain records created-but-not-yet-ready-to-synchronize, which also sounds like a legit improvement to Watermelon Sync that could be achieved by formalizing a new `draft` sync status without a generic customization point.
> 
> WDYT?

exact this is what i need =)

File: src/sync/impl/fetchLocal.js [hunk 2]
```diff
@@ -11,45 +11,77 @@ import {
...
+async function fetchLocalChangesForCollection<T: Model>({
+  collection,
+  createdQueries = [],
+  updatedQueries = [],
+  deletedQueries = [],
+}: fetchLocalChangesForCollectionArgs<T>): Promise<[SyncTableChangeSet, T[]]> {
   const [createdRecords, updatedRecords, deletedRecords] = await Promise.all([
-    collection.query(createdQuery).fetch(),
-    collection.query(updatedQuery).fetch(),
-    collection.database.adapter.getDeletedRecords(collection.table),
```
[the remark this answers is not in the document]
Comment (sidferreira) [reply src/sync/impl/fetchLocal.js:28]: Yes, I totally agree with you on this point and that's why I mentioned `It required more changes than I expected and I'm not sure I'm happy with it.`. 

At the same time, I think we should have options to query deleted especially to avoid `collection.database.adapter.getDeletedRecords(collection.table)`, that basically breaks the `model => collection => database` structure (part of the decoupling I mentioned).

Maybe changing to `collection.query(updatedQuery).fetchIds()` to be a very specific method.

I'm not happy with this solution, especially on the part that skips the cache, but I think there are too many hard-coded rules (`createdQuery`, `updatedQuery`) that make sense to be customizable IMHO.

In short, I'm not happy with the purposed solution, but I think it's worth thinking about the concept of customization.

Even the `draft` state might be troublesome, as some users may want the `drafts` to be synced. An email app would like `drafts` to be synced. And in this case, they would need another `status` field.

I believe that #1027 starts a discussion about `pullable, pushable, local-only, read-only` scenarios.

**Comment by sidferreira** [comment]
@radex btw, I don't have the needs that @osvaldokalvaitir described, I just purposed something based on what I felt when I put on his shoes.

`draft` does solve this issue, but maybe we can think a bit more broadly and have more options.

File: src/Collection/index.js (hunk 1 above)
Comment (sidferreira) [reply src/Collection/index.js:86]: @radex undelete?!

**Comment by radex** [comment]
@sidferreira hey, any reason why you closed this and another PR of yours?

**Comment by sidferreira** [comment]
@radex it was marked as won't fix... I just assumed it wasn't relevant...

**Comment by radex** [comment]
no, no, it's was just a rogue bot - I appreciate your PRs, and while I can't get to properly reviewing them now, they are useful
